### PR TITLE
One hot encoding

### LINKED ISSUE #81: One Hot Encoding
### Description

Currently, DeepMol lacks a SMILES one hot encoder, which is a commonly used representation for molecular structures. This issue proposes the addition of a SMILES one hot encoder to the package.

### Feature Request

Add a SMILES one hot encoder to DeepMol that allows users to convert SMILES strings into one hot encoded vectors, which can be used for further analysis and modeling in cheminformatics.

### Proposed Solution

The SMILES one hot encoder should be implemented as a new class inheriting from `MolecularFeaturizer`. The encoder should take a `Dataset` as input and return the same dataset with one hot encoded vectors representing the molecular structures as the `X` property. Additionally, the encoder should support handling various aspects of SMILES notation, such as chirality and aromaticity, as these can affect the representation of the molecules.

### PR DESCRIPTION
**Pull request by jcorreia11**
### Issue Description:

Currently, DeepMol lacks a SMILES one hot encoder, which is a commonly used representation for molecular structures. This pull request aims to address this issue by proposing the addition of a SMILES one hot encoder to the package.

### Solution:

The proposed implementation one hot encodes SMILES strings. 

Key features:
- Two letter atoms are considered as one token (e.g. Cl, Br, Na, Cl, etc)
- It is possible to use regex (by default true) to encode '[\*]' as one token \*
- Added the abstract class Tokenizer. Child classes must implement the encode and decode methods and the shape property.

 \* <sup><sub>Atoms are represented by their atomic symbols: this is the only required use of letters in SMILES. Each non-hydrogen atom is specified independently by its atomic symbol enclosed in square brackets, [ ]. The second letter of two-character symbols must be entered in lowercase. Elements in the "organic subset" B, C, N, O, P, S, F, Cl, Br, and I may be written without brackets if the number of attached hydrogens conforms to the lowest normal valence consistent with explicit bonds.<sup><sub>

### Usage

```python
from deepmol.tokenizers import SmilesOneHotEncoder
from deepmol.loaders import CSVLoader

data = loader = CSVLoader('data_path.csv', smiles_field='Smiles', labels_fields=['Class'])
dataset = loader.create_dataset(sep=";")
ohe = SmilesOneHotEncoder().fit_transform(dataset)  # or .featurize
```

Closes #81 .

### CODE REVIEW

**Comment by jcorreia11** [comment]
@jcapels, how do you think we should deal with molecules with length higher than the provided? Remove them, one-hot-encode smiles[:max_length] or in other way?

**Comment by jcapels** [comment]
Hi, 

I use to truncate as you mentioned (smiles[:max_length]). 

Did you consider to include SmilesPE? It does the work at the atom-level and substructure level. https://pypi.org/project/SmilesPE/.

**Review by jcapels — APPROVED** [review]
Seems good, see the comment above about the tokenizer and truncation

**Comment by jcorreia11** [comment]
@jcapels  I was not aware of that package... 
I will take a look at it!

**Comment by jcorreia11** [comment]
Hi @jcapels. I just made some significant changes in the PR. Now we have a module for Tokenizers (with atom-level and kmers tokenizers already implemented for smiles strings (later we can decide if it is worth adding others)). The one hot encoding is now part of the `compound_featurization` module. The `SmilesOneHotEncoder` is working well with both tokenizers.

When you have time please take a new look at the PR.

Some notes:

- The `tokenizers` inherit from the `Estimator` class not from the `Transformer` (it does not implement the transform method, instead implements a tokenize method that returns the tokens instead of a transformed dataset).
- For smiles with a length greater than `max_length`, one hot encoding is done for the first part of the smiles (`smiles[:max_length]`). Should we add an argument to allow customization of how this is done (one hot encode initial, final, medium part, or remove)?
- Padding is done by adding the `""` (empty) char to the dictionary. However, the way padding is done could be improved to also allow more customization.
- The regex used for atom-level tokenization (`"(\[[^\]]+]|Br?|Cl?|N|O|S|P|F|I|b|c|n|o|s|p|\(|\)|\.|=|#|-|\+|\\\\|\/|:|~|@|\?|>|\*|\$|\%[0-9]{2}|[0-9])"`) was taken from:
[1]  Philippe Schwaller, Teodoro Laino, Théophile Gaudin, Peter Bolgar, Christopher A. Hunter, Costas Bekas, and Alpha A. Lee
ACS Central Science 2019 5 (9): Molecular Transformer: A Model for Uncertainty-Calibrated Chemical Reaction Prediction
1572-1583 DOI: 10.1021/acscentsci.9b00576

It is a reliable source so it should cover the majority of cases, however, I did not extensively tested for that.

### Future work

It could be interesting to add pre-trained tokenizers (however those generally require heavy computational resources).

I will leave here some relevant links for future reference:

https://github.com/XinhaoLi74/SmilesPE/tree/master#use-the-pre-trained-smilespe-tokenizer

https://huggingface.co/DeepChem/SmilesTokenizer_PubChem_1M/tree/main

https://github.com/deepchem/deepchem/blob/master/deepchem/feat/smiles_tokenizer.py#L37

https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer

https://huggingface.co/seyonec/PubChem10M_SMILES_BPE_450k

**Review by jcapels — APPROVED** [review]
Approved
